Reload NSpec service

lchanmann opened the pull request:
Reset service factory after the service setting is changed so that the nspec service is updated accordingly.

Addresses . https://nanopay.atlassian.net/browse/CPF-2149

lchanmann committed:
Reset service factory after the service setting is changed

kgrgreer commented on src/foam/core/SingletonFactory.java:
```diff
@@ -19,10 +19,14 @@ public SingletonFactory(XFactory delegate) {
     return instance_;
   }
+
+  public void setDelegate(XFactory delegate) {
```
Needs to be synchronized.

lchanmann replied:
Fixed.

kgrgreer commented on src/foam/nanos/boot/Boot.java:
```diff
@@ -58,25 +61,21 @@ public Boot(String datadir) {
 
     for ( int i = 0 ; i < l.size() ; i++ ) {
       NSpec sp = (NSpec) l.get(i);
+      XFactory factory = new SingletonFactory(new NSpecFactory((ProxyX) root_, sp));
```
Why do we need to decorate in a SingletonFactory? Doesn't NSpecFactory already cache the service?

kgrgreer replied:
Just add an invalidate() method on the NSpecFactory and call it instead.

lchanmann replied:
Fixed.

kgrgreer commented on src/foam/nanos/boot/Boot.java:
```diff
@@ -58,25 +61,21 @@ public Boot(String datadir) {
-          }
-        }
+        logger.info("Reload service factory:", sp.getName());
+        ((SingletonFactory) factories_.get(sp.getName()))
```
Cast to an NSpecFactory and call invalidate() instead.

lchanmann replied:
Fixed.

lchanmann committed:
Merge branch master into ReloadNSpecService

lchanmann committed:
Invalidate NSpecFactory of the updated service

instead of decorating with SingletonFactory since NSpecFactory already supports caching.

lchanmann committed:
Conditional NSpec invalidation

lchanmann committed:
Reload web agent and nano service for the updated NSpec

lchanmann committed:
Synchronize SingletonFactory.setDelegate()

lchanmann commented:
Addressed your comments. @kgrgreer

kgrgreer commented on src/foam/nanos/boot/Boot.java:
```diff
@@ -13,23 +13,25 @@
 public class Boot {
   // Context key used to store the top-level root context in the context.
   public final static String ROOT = "_ROOT_";
+  private final static Set<String> RELOADED_SERVICES = new HashSet<>();
```
Why is this static? This would prevent me from booting two NANOS servers in the same VM. Can't it just be an instance variable?

lchanmann replied:
It needs a way to tell if a service is reloaded so that the already loaded web agents and nano services can be reloaded - https://github.com/foam-framework/foam2/pull/2423/commits/9649657c7c8f74325a77ce9d6ac498133088a18d.

> Why is this static? This would prevent me from booting two NANOS servers in the same VM.

It is not an instance variable because we do not pass the instance of Boot to NanoRouter/NanoServiceRouter.

I'm open to change. What are the other alternative you have in mind?

lchanmann replied:
Changed to `reloadedServices_` instance variable and store it in the context to be accessible in NanoRouter and NanoServiceRouter- https://github.com/foam-framework/foam2/pull/2423/commits/9a7e40c31279e26dad07658ca1101d65125d5150.

kgrgreer commented on src/foam/nanos/boot/Boot.java:
```diff
@@ -13,23 +13,25 @@
 public class Boot {
   // Context key used to store the top-level root context in the context.
   public final static String ROOT = "_ROOT_";
+  private final static Set<String> RELOADED_SERVICES = new HashSet<>();
```
What is this Set used for?

lchanmann replied:
It's for storing the reloaded service name for use in reloading web agents and nano services.

lchanmann committed:
Put reloaded services in the context instead of storing as a static field

lchanmann commented:
Addressed your comments. @kgrgreer Please take another look, thanks.

jlhughes commented:
@lchanmann can you explain the reloadedServices_ , I'm still not clear on its purpose reading through the code.

lchanmann commented:
We need a way to communicate with NanoRouter/NanoServiceRouter that a web agent or nano service needs to be reloaded since these routers also have a cached version of it - https://github.com/foam-framework/foam2/blob/master/src/foam/nanos/http/NanoRouter.java#L109.

So reloadedServices_ would store the service name that was invalidated - https://github.com/lchanmann/foam2/blob/ReloadNSpecService/src/foam/nanos/boot/Boot.java#L78 and NanoRouter/NanoServiceRouter can act accordingly.

jlhughes commented:
@lchanmann got it, thanks.

kgrgreer commented on src/foam/nanos/boot/Boot.java:
```diff
@@ -13,23 +13,26 @@
 public class Boot {
   // Context key used to store the top-level root context in the context.
   public final static String ROOT = "_ROOT_";
+  public final static String RELOADED_SERVICES = "_RELOADED_SERVICES_";
```
Have Nanorouter listen to changes in the NSpecDAO and get rid of this.

lchanmann replied:
Fixed in https://github.com/foam-framework/foam2/pull/2423/commits/e5fed73a234dbc4a10fdf6540fb95a181405f07a.

lchanmann committed:
Listen for changes in nSpecDAO in NanoRouter

lchanmann committed:
Eager reload lazy:false services

lchanmann committed:
Move eager reload logic to NSpecFactory.invalidate

lchanmann committed:
Optimize imports

lchanmann commented:
Addressed your comments. @kgrgreer

lchanmann commented:
There is a [NanoServiceRouter.java](https://github.com/foam-framework/foam2/blob/master/src/foam/nanos/box/NanoServiceRouter.java) but I'm not quite sure when it's being used so leaving it unchanged.
